### PR TITLE
#20: Support @path on controllers methods (closes #20)

### LINKED ISSUE #20: Support @path on controllers methods
## requirements
Support the use of `@path` on controllers' methods to override the uri path name with the following semantics:

```scala
@path("foo")
trait FooController {

  @query @path("baz")
  def bar(): Bar

  @query @path("bar/search")  // we agreed not to implement for the moment
  def barSearch(): Bar 

}
```

publishes the following routes:

```http
GET /foo/baz
GET /foo/bar/search #we agreed not to implement for the moment
```

### PR DESCRIPTION
**Pull request by calippo**
Closes #20

### tests performed
- tested locally
- added unit tests

### CODE REVIEW

File: core/src/main/scala/annotations.scala [hunk 1]
```diff
@@ -40,17 +50,26 @@ package object annotation {
...
-          q"$annots def $methodName(...$allArgs): $tpe = $body"
+          val operationName = c.universe.TermName(chooseOperationName(
+            c = c,
+            methodName = methodName.toString
+          ))
+          q"$annots def $operationName(...$allArgs): $tpe = $body"
         case q"$annots def $methodName(...$args): $tpe" :: nil =>
           val allArgs = args.map{
             case innerArgs => q"val actionCommand: String" +: innerArgs
           }
```
Comment (bytecodeguru) [inline core/src/main/scala/annotations.scala]: Let's try to improve readability by introducing a local method `treeFrom(args: ...): Tree`

File: core/src/main/scala/annotations.scala [hunk 2]
```diff
@@ -40,17 +50,26 @@ package object annotation {
...
     def impl(c: Context)(annottees: c.Expr[Any]*): c.Expr[Any] = {
       import c.universe._
+
       val result = annottees.map(_.tree).toList match {
         case q"$annots def $methodName(...$args): $tpe = $body" :: nil =>
           val allArgs = args.map{
             case innerArgs => q"val actionCommand: String" +: innerArgs
           }
-          q"$annots def $methodName(...$allArgs): $tpe = $body"
+          val operationName = c.universe.TermName(chooseOperationName(
```
Comment (bytecodeguru) [inline core/src/main/scala/annotations.scala]: In this case I'd prefer this style of indentation:
```
val operationName = c.universe.TermName(
  chooseOperationName(c, methodName.toString)
)
```

**Review by bytecodeguru — CHANGES_REQUESTED** [review]
It's OK for me, just my two cents about code readability

File: core/src/main/scala/annotations.scala (hunk 2 above)
Comment (calippo) [reply core/src/main/scala/annotations.scala]: +1

Though I like named arguments

Comment (bytecodeguru) [reply core/src/main/scala/annotations.scala]: Yeah so do I, but in this specific case...

File: core/src/main/scala/annotations.scala (hunk 1 above)
Comment (calippo) [reply core/src/main/scala/annotations.scala]: I've already been thinking about how to improve the readability of this module, and it's kind of tricky.

- Quasiquotes are pretty dense, I'm not sure introducing a method would make code more readable (the signature would be awful)
- It not that easy to generalize the method for both `command` and `query` (the use different `c.universe`)

Comment (calippo) [reply core/src/main/scala/annotations.scala]: Well, I thought you referred to the line above. What you propose it's actually very easy :)

File: core/src/main/scala/annotations.scala [hunk 3]
```diff
@@ -51,25 +50,21 @@ package object annotation {
...
         case q"$annots def $methodName(...$args): $tpe = $body" :: nil =>
-          val allArgs = args.map{
-            case innerArgs => q"val actionCommand: String" +: innerArgs
-          }
-          val operationName = c.universe.TermName(chooseOperationName(
-            c = c,
-            methodName = methodName.toString
-          ))
-          q"$annots def $operationName(...$allArgs): $tpe = $body"
+          val commandArgs = args.map(commandValDef +: _)
```
Comment (calippo) [inline core/src/main/scala/annotations.scala]: @bytecodeguru I ended up with this

Comment (bytecodeguru) [reply core/src/main/scala/annotations.scala]: 👍

**Review by bytecodeguru — APPROVED** [review]

**Pull request merged**